gobblin-core/src/main/java/gobblin/converter/csv/CsvToJsonConverterV2.java

```diff
@@ -0,0 +1,169 @@
+    if (!customOrder.isEmpty()) {
+      LOG.info("Will use custom order to generate JSON from CSV: " + customOrder);
+    }
+    return this;
+  }
+  @Override
+  public JsonArray convertSchema(String inputSchema, WorkUnitState workUnit) throws SchemaConversionException {
+    Preconditions.checkNotNull(inputSchema, "inputSchema is required.");
+    return JSON_PARSER.parse(inputSchema).getAsJsonArray();
+  }
+
+  /**
+   * Converts CSV (array of String) to JSON.
+   * By default, fields between CSV and JSON are mapped in order bases and it validates if both input and output has same number of fields.
+   *
+   * Customization can be achieved by adding custom order where user can define list of indices of CSV fields correspond to JSON schema user defined.
+   *
+   * e.g:
+   * - CSV
+   *   "Street number", "City" , "Country", "Zip code"
+   *
+   * - Custom order
+   *   0,1,3
+   * - Output JSON (Key is derived from input schema)
+   *   {"street_number" : "Street number", "city" : "City" , "zip_code" : "Zip code" }
+   *
+   * - Custom order (adding null when negative index is defined)
+   *   0,1,-1,3
+   * - Output JSON
+   *   {"street_number" : "Street number", "city" : "City" , "Country" : null, "zip_code" : "Zip code" }
+   *
+   *
+   * {@inheritDoc}
+   * @see gobblin.converter.Converter#convertRecord(java.lang.Object, java.lang.Object, gobblin.configuration.WorkUnitState)
+   */
+  @Override
+  public Iterable<JsonObject> convertRecord(JsonArray outputSchema, String[] inputRecord, WorkUnitState workUnit)
+      throws DataConversionException {
+
+    JsonObject outputRecord = null;
+    if (!customOrder.isEmpty()) {
+      outputRecord = createOutput(outputSchema, inputRecord, customOrder);
+    } else {
+      outputRecord = createOutput(outputSchema, inputRecord);
+    }
+
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Converted into " + outputRecord);
+    }
+    return new SingleRecordIterable<JsonObject>(outputRecord);
+  }
+
+  @VisibleForTesting
+  JsonObject createOutput(JsonArray outputSchema, String[] inputRecord) {
+    Preconditions.checkArgument(outputSchema.size() == inputRecord.length, "# of columns mismatch. Input "
+        + inputRecord.length + " , output: " + outputSchema.size());
+    JsonObject outputRecord = new JsonObject();
+
+    for (int i = 0; i < outputSchema.size(); i++) {
+      String key = outputSchema.get(i).getAsJsonObject().get(COLUMN_NAME_KEY).getAsString();
+
+      if (StringUtils.isEmpty(inputRecord[i]) || JSON_NULL_VAL.equalsIgnoreCase(inputRecord[i])) {
+        outputRecord.add(key, JsonNull.INSTANCE);
+      } else {
+        outputRecord.addProperty(key, inputRecord[i]);
+      }
+    }
+
+    return outputRecord;
+  }
+
+  @VisibleForTesting
+  JsonObject createOutput(JsonArray outputSchema, String[] inputRecord, List<String> customOrder) {
+
+    Preconditions.checkArgument(outputSchema.size() == customOrder.size(), "# of columns mismatch. Input "
```

Comment:
As per the example: customOrder can be lesser than outputSchema size. Is my understanding correct? If so this check should be changed